designs/webviews/webviews-general.md

```diff
@@ -0,0 +1,95 @@
+| Deploy SAM Application               | Yes              | Would be a more logical flow when populating data based on region. Users can change region and invalidate other fields, and refresh potential selections. |
+
+VS Code webviews are not a cureall, and introduce a new set of complexities. A list, along with mitigations, follows:
+
+* Webviews are HTML + JS running in an `iframe`, which adds a considerable resource burden on the IDE (see Appendix 1).
+  * Webviews offer a `retainContextWhenHidden` which is false by default. If false, when the webview goes to the background, it is no longer rendered, and re-rendered when it regains context.
+* If webviews aren't configured to persist context when a user moves it to the background, the webview will have to re-render, losing any unsaved state.
+  * This is only a concern if `retainContextWhenHidden` is false (which is recommended). VS Code lets you set a JSON-serilaizable state which can be reloaded when regaining context, via prorpietary VS Code functions, `vscode.getState` and `vscode.setState`.
+    * This can also be used to restore state between VS Code sessions using a `WebviewPanelSerializer`.
+* Webviews are separate from the background process.
+  * Webviews feature a two-way messaging system, which are by default untyped. All messages must be valid JSON.
+    * The backend can post via `vscode.WebviewPanel.webview.postMessage`, which can then be picked up by an `eventListener` on the frontend, listening for  `message` events.
+      * One caveat: messages should only be posted to the webview while the webview is active, ideally on frontend request.
+        * Messages to dormant webviews are batched and received by the webview on load, but are added to the Javascript event queue. If the webview is rehibernated prior to processing the message from the queue, the message will be lost as VS Code will have sent it, but the event queue will have been wiped.
+        * Messages should additionally carry as full of a state as possible; this makes webviews more resilient to dropping messages (since the next state will carry a full slate of information)
+    * The frontend can post to the backend by proprietary VS Code function, `vscode.postMessage`, which is then picked up by the frontend via `vscode.WebviewPanel.webview.onDidReceiveMessage`, which invokes a callback function on the message.
+
+## Appendicies
+
+### Appendix 1: "Should I use a webview?" -- from the [VS Code documentation](https://code.visualstudio.com/api/extension-guides/webview#should-i-use-a-webview)
+
+Webviews are pretty amazing, but they should also be used sparingly and only when VS Code's native API is inadequate. Webviews are resource heavy and run in a separate context from normal extensions. A poorly designed webview can also easily feel out of place within VS Code.
```

Comment:
Consider markdown quoting this entire section since its contents are from the other page.